extensions/arc/notebooks/arc/content/postgres/tsg100-troubleshoot-postgres.ipynb

```diff
@@ -0,0 +1,590 @@
+    "        # On Mac, when ADS is not launched from terminal, LC_ALL may not be set, which causes pip installs to fail\n",
+    "        # with:\n",
+    "        #\n",
+    "        #    UnicodeDecodeError: 'ascii' codec can't decode byte 0xc5 in position 4969: ordinal not in range(128)\n",
+    "        #\n",
+    "        # Setting it to a default value of \"en_US.UTF-8\" enables pip install to complete\n",
+    "        #\n",
+    "        if platform.system() == \"Darwin\" and \"LC_ALL\" not in os.environ:\n",
+    "            os.environ[\"LC_ALL\"] = \"en_US.UTF-8\"\n",
+    "\n",
+    "    # When running `kubectl`, if AZDATA_OPENSHIFT is set, use `oc`\n",
+    "    #\n",
+    "    if cmd.startswith(\"kubectl \") and \"AZDATA_OPENSHIFT\" in os.environ:\n",
+    "        cmd_actual[0] = cmd_actual[0].replace(\"kubectl\", \"oc\")\n",
+    "\n",
+    "    # To aid supportabilty, determine which binary file will actually be executed on the machine\n",
+    "    #\n",
+    "    which_binary = None\n",
+    "\n",
+    "    # Special case for CURL on Windows.  The version of CURL in Windows System32 does not work to\n",
+    "    # get JWT tokens, it returns \"(56) Failure when receiving data from the peer\".  If another instance\n",
+    "    # of CURL exists on the machine use that one.  (Unfortunately the curl.exe in System32 is almost\n",
+    "    # always the first curl.exe in the path, and it can't be uninstalled from System32, so here we\n",
+    "    # look for the 2nd installation of CURL in the path)\n",
+    "    if platform.system() == \"Windows\" and cmd.startswith(\"curl \"):\n",
+    "        path = os.getenv('PATH')\n",
+    "        for p in path.split(os.path.pathsep):\n",
+    "            p = os.path.join(p, \"curl.exe\")\n",
+    "            if os.path.exists(p) and os.access(p, os.X_OK):\n",
+    "                if p.lower().find(\"system32\") == -1:\n",
+    "                    cmd_actual[0] = p\n",
+    "                    which_binary = p\n",
+    "                    break\n",
+    "\n",
+    "    # Find the path based location (shutil.which) of the executable that will be run (and display it to aid supportability), this\n",
+    "    # seems to be required for .msi installs of azdata.cmd/az.cmd.  (otherwise Popen returns FileNotFound) \n",
+    "    #\n",
+    "    # NOTE: Bash needs cmd to be the list of the space separated values hence shlex.split.\n",
+    "    #\n",
+    "    if which_binary == None:\n",
+    "        which_binary = shutil.which(cmd_actual[0])\n",
+    "\n",
+    "    if which_binary == None:\n",
+    "        if user_provided_exe_name in install_hint and install_hint[user_provided_exe_name] is not None:\n",
+    "            display(Markdown(f'HINT: Use [{install_hint[user_provided_exe_name][0]}]({install_hint[user_provided_exe_name][1]}) to resolve this issue.'))\n",
+    "\n",
+    "        raise FileNotFoundError(f\"Executable '{cmd_actual[0]}' not found in path (where/which)\")\n",
```

Comment:
what does where/which refer to here?

Reply:
Presumably they refer to the programs where (Windows) and which (Linux) that look up where executables are located.  Btw this run function is referenced from an existing snippet that I didn't author. (projects/adsbooks/common-all/.snippets/def run().md in the aris repo)